### PR TITLE
Add keypoints matching visualization for real-time pose estimation tutorial

### PR DESCRIPTION
**Pull request by catree**
- Lots of auto formatting code changes.
- Fix parsing of arguments:
https://github.com/opencv/opencv/blob/54a1bbe1c42ceeb3bd0875e7c95ffcc0d883f1e4/samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/main_detection.cpp#L105-L112
- Add keypoints matching visualization (only after training a new model file)
- Add possibility to use another type of keypoints (not only ORB)

---

It is not possible to get the same detection results with the provided image and video than in the tutorial video.

Beware of the Kalman filter formulation used (constant acceleration on Euler angles). Here what looks like a better [formulation](https://github.com/chili-epfl/chilitags/pull/75).

---

Some video results (unfortunately, YouTube applied too much compression):

Training pose:
```
cv::Mat rvec = (cv::Mat_<double>(3,1) << 2.073596928195101, -0.8861248486736736, 0.4648378262984375);
cv::Mat tvec = (cv::Mat_<double>(3,1) << -65.34756669790684, -34.34732391473651, 361.0313725421983);
```

- ORB: https://youtu.be/blI5xMzLgVU
- SURF: https://youtu.be/F2UVIKIN-vg
- SIFT: https://youtu.be/AM6cCFEoaow
- BRISK: https://youtu.be/Eq2WNv1blfc
- KAZE: https://youtu.be/cbiFDMV4XBM
- AKAZE: https://youtu.be/ug0SfTLaf6w
- BINBOOST: https://youtu.be/EnbrVD1iWEY
- VGG: https://youtu.be/wBTYLSvR3ZE

### CODE REVIEW

File: samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/Model.cpp [hunk 1]
```diff
@@ -8,66 +8,76 @@
...
 
-  cv::FileStorage storage(path, cv::FileStorage::WRITE);
-  storage << "points_3d" << points3dmatrix;
-  storage << "points_2d" << points2dmatrix;
-  storage << "keypoints" << list_keypoints_;
-  storage << "descriptors" << descriptors_;
+    cv::FileStorage storage(path, cv::FileStorage::WRITE);
+    storage << "points_3d" << points3dmatrix;
+    storage << "points_2d" << points2dmatrix;
+    storage << "keypoints" << list_keypoints_;
```
Comment (catree) [inline samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/Model.cpp]: Modification to save keypoints and path to the training image. This way, during the detection it is possible to display the keypoints matching.

File: samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/Model.cpp [hunk 2]
```diff
@@ -8,66 +8,76 @@
...
-  points3d_mat.copyTo(list_points3d_in_);
-
-  storage.release();
-
+    cv::Mat points3d_mat;
+
+    cv::FileStorage storage(path, cv::FileStorage::READ);
+    storage["points_3d"] >> points3d_mat;
+    storage["descriptors"] >> descriptors_;
+    if (!storage["keypoints"].empty())
```
Comment (catree) [inline samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/Model.cpp]: Read keypoints and path to training image if node is not empty.

File: samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/PnPProblem.h [hunk 3]
```diff
@@ -18,41 +18,35 @@
...
+    /** The computed rotation matrix */
+    cv::Mat R_matrix_;
+    /** The computed translation matrix */
+    cv::Mat t_matrix_;
+    /** The computed projection matrix */
+    cv::Mat P_matrix_;
 };
 
-// Functions for Möller-Trumbore intersection algorithm
-cv::Point3f CROSS(cv::Point3f v1, cv::Point3f v2);
```
Comment (catree) [inline samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/PnPProblem.h]: Functions declaration not needed here.

File: samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/Utils.cpp [hunk 4]
```diff
@@ -11,178 +11,180 @@
...
-  cv::Scalar blue(255,0,0);
-  cv::Scalar black(0,0,0);
-
-  cv::Point2i origin = list_points2d[0];
-  cv::Point2i pointX = list_points2d[1];
-  cv::Point2i pointY = list_points2d[2];
-  cv::Point2i pointZ = list_points2d[3];
-
-  drawArrow(image, origin, pointX, red, 9, 2);
-  drawArrow(image, origin, pointY, blue, 9, 2);
```
Comment (catree) [inline samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/Utils.cpp]: Fix Y axis drawn in blue (should be green).

File: samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/Utils.cpp [hunk 5]
```diff
@@ -234,65 +236,166 @@ cv::Mat rot2euler(const cv::Mat & rotationMatrix)
...
 {
-  std::ostringstream ss;
-  ss << Number;
-  return ss.str();
+    std::ostringstream ss;
+    ss << Number;
+    return ss.str();
+}
+
+void createFeatures(const std::string &featureName, int numKeypoints, cv::Ptr<cv::Feature2D> &detector, cv::Ptr<cv::Feature2D> &descriptor)
```
Comment (catree) [inline samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/Utils.cpp]: Add possibility to use different types of features.

File: samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/main_detection.cpp [hunk 6]
```diff
@@ -21,451 +20,483 @@
...
+
+    const String keys =
+            "{help h            |      | print this message                                                 }"
+            "{video v           |      | path to recorded video                                             }"
+            "{model             |      | path to yml model                                                  }"
+            "{mesh              |      | path to ply mesh                                                   }"
+            "{keypoints k       |2000  | number of keypoints to detect                                      }"
+            "{ratio r           |0.7   | threshold for ratio test                                           }"
+            "{iterations it     |500   | RANSAC maximum iterations count                                    }"
+            "{error e           |6.0   | RANSAC reprojection error                                          }"
```
Comment (catree) [inline samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/main_detection.cpp]: Increase RANSAC reprojection (2 pixel error seems too harsh, some valid points could be treated as outliers, especially with the incertitude errors about the intrinsic camera parameters and the camera distortion) threshold and RANSAC confidence.

**Comment by catree** [comment]
I have put some comments to highlight the relevant changes since lots of the changes are whitespace related.

File: samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/main_registration.cpp [hunk 7]
```diff
@@ -58,211 +46,250 @@ Model model;
...
+            "{keypoints k   |2000  | number of keypoints to detect (only for ORB)                       }"
+            "{feature       |ORB   | feature name (ORB, KAZE, AKAZE, BRISK, SIFT, SURF, BINBOOST, VGG)  }"
+            ;
+    CommandLineParser parser(argc, argv, keys);
+
+    string tutorial_path = "../../samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/"; // path to tutorial
+
+    string img_path = tutorial_path + "Data/resized_IMG_3875.JPG";  // image to register
+    string ply_read_path = tutorial_path + "Data/box.ply";          // object mesh
+    string write_path = tutorial_path + "Data/cookies_ORB.yml";     // output file
```
Comment (alalek) [inline samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/main_registration.cpp]: Could you try to use `samples::findFile()` from here: https://github.com/opencv/opencv/pull/12354 ?
Perhaps "../../samples" prefix should be dropped.

Comment (catree) [reply samples/cpp/tutorial_code/calib3d/real_time_pose_estimation/src/main_registration.cpp]: This works with the updated changes (on Ubuntu).

**Pull request merged**